datalad/cmd.py

```diff
@@ -98,7 +98,208 @@ def _cleanup_output(stream, std):
         std.close()
 
 
-class Runner(object):
+def kill_output(output):
+    """Helper for WitlessRunner to swallow all output and neither
+    relay it to the parent process's stdout, not provide it as
+    return value of WitlessRunner.run().
+    """
+    return b'', 0
+
+
+def capture_output(output):
+    """Helper for WitlessRunner to capture all output and
+    provide it as return value of WitlessRunner.run().
+    """
+    return output, 0
+
+
+class WitlessRunner(object):
+    """Minimal Runner with support for online command output processing
+
+    It aims to be as simple as possible, providing only essential
+    functionality. Derived classes should be used for additional
+    specializations and convenience features.
+    """
+    __slots__ = ['cwd', 'env', '_poll_period']
+
+    def __init__(self, cwd=None, env=None, poll_period=0.1):
+        """
+        Parameters
+        ----------
+        cwd : path-like, optional
+          If given, commands are executed with this path as PWD,
+          the PWD of the parent process is used otherwise.
+        env : dict, optional
+          Environment to be pass to subprocess.Popen(). If `cwd`
+          was given, 'PWD' in the environment is set to its value.
+          This must be a complete environment definition, no values
+          from the current environment will be inherited.
+        poll_period : float, optional
+          Interval at which the running process is queried for
+          output.
+        """
+        self._poll_period = poll_period
+        self.env = env.copy() if env else None
+        # stringify to support Path instances on PY35
+        self.cwd = str(cwd) if cwd is not None else None
+        if self.cwd and env is not None:
+            # if CWD was provided, we must not make it conflict with
+            # a potential PWD setting
+            self.env['PWD'] = self.cwd
+
+    def run(self, cmd, proc_stdout=None, proc_stderr=None, stdin=None):
+        """Execute a command and communicate with it.
+
+        Parameters
+        ----------
+        cmd : list
+          Sequence of program arguments. Passing a single string means
+          that it is simply the name of the program, no complex shell
+          commands are supported.
+        proc_stdout : callable, optional
+          By defaultno stdout is captured, but relayed to the parent
+          process's stdout. If given, all stdout is passed as byte-string
+          to this callable, in the chunks it was received by polling the
+          processing. The callable may transform it in any way, its output
+          (byte-string) is concatenated and provided as stdout return value.
+          The helper functions 'kill_output', and 'capture_output' are
+          provided to either swallow all output (and not relay it to the
+          parent), or to capture all output and provide it as return
+          value.
```

Comment:
needs to expand on returned value to be a tuple of the `(processed, unprocessed_len)`  with the description of what `unprocessed_len` is (I guess the length of the trailing portion of the input value provided to the callable, which is yet to be processed.  but could be also the trailing portion within returned value which is yet to be processed)

Reply:
True, will do. Thx.

Reply:
Done